zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/AuthenticationHandler.java

```diff
@@ -0,0 +1,127 @@
+   *
+   * @return either a {@link Status} if the authentication failed, or a {@link Context} with
+   *     authentication information if the authentication succeeded.
+   */
+  Either<Status, Context> authenticate(String authorizationHeader);
+
+  final class Oidc implements AuthenticationHandler {
+    public static final Context.Key<Map<String, Object>> USER_CLAIMS =
+        Context.key("io.camunda.zeebe:user_claim");
+
+    public static final String BEARER_PREFIX = "Bearer ";
+    private final JwtDecoder jwtDecoder;
+
+    public Oidc(final JwtDecoder jwtDecoder) {
+      this.jwtDecoder = Objects.requireNonNull(jwtDecoder);
+    }
+
+    @Override
+    public Either<Status, Context> authenticate(final String authorizationHeader) {
+      if (!authorizationHeader.startsWith(BEARER_PREFIX)) {
+        return Either.left(
+            Status.UNAUTHENTICATED.augmentDescription(
+                "Expected authentication information to start with '%s'".formatted(BEARER_PREFIX)));
+      }
+
+      final Jwt token;
+      try {
+        token = jwtDecoder.decode(authorizationHeader.substring(BEARER_PREFIX.length()));
+      } catch (final JwtException e) {
+        return Either.left(
+            Status.UNAUTHENTICATED
+                .augmentDescription("Expected a valid token, see cause for details")
+                .withCause(e));
+      }
+
+      return Either.right(Context.current().withValue(USER_CLAIMS, token.getClaims()));
+    }
+  }
+
+  final class BasicAuth implements AuthenticationHandler {
+    public static final Context.Key<String> USERNAME = Context.key("io.camunda.zeebe:username");
+    private static final String BASIC_PREFIX = "Basic ";
+    private final UserServices userServices;
+    private final PasswordEncoder passwordEncoder;
+
+    public BasicAuth(final UserServices userServices, final PasswordEncoder passwordEncoder) {
+      this.userServices = userServices;
+      this.passwordEncoder = passwordEncoder;
+    }
+
+    @Override
+    public Either<Status, Context> authenticate(final String authorizationHeader) {
+      if (!authorizationHeader.startsWith(BASIC_PREFIX)) {
+        return Either.left(
+            Status.UNAUTHENTICATED.augmentDescription(
+                "Expected authentication information to start with '%s'".formatted(BASIC_PREFIX)));
+      }
+
+      final var decodedAuth =
+          new String(
+              Base64.getDecoder().decode(authorizationHeader.substring(BASIC_PREFIX.length())));
+      final var authParts = decodedAuth.split(":", 2);
+      final var username = authParts[0];
+      final var password = authParts[1];
+
+      final Optional<UserEntity> userOpt;
+      try {
+        userOpt = loadUserByUsername(username);
+      } catch (final RuntimeException e) {
+        return Either.left(
+            Status.UNAUTHENTICATED.augmentDescription("Failed to authenticate").withCause(e));
+      }
+
+      if (userOpt.isEmpty()) {
```

Comment:
💭 I wondered as the status and the description are the same could we condense this and the if check below into one statement? 

Ultimately its preference and NABD so I'm not going to push :)

Reply:
Yeah maybe, but let's leave it as is. Especially for security relevant code, I like the KISS principle.